src/passes/Monomorphize.cpp

```diff
@@ -0,0 +1,197 @@
+        if (call->type == Type::unreachable) {
+          // Ignore unreachable code.
+          // TODO: call_return?
+          continue;
+        }
+
+        if (call->target == name) {
+          // Avoid recursion, which adds some complexity (as we'd be modifying
+          // ourselves if we apply optimizations).
+          continue;
+        }
+
+        call->target = getRefinedTarget(call, module);
+      }
+    }
+  }
+
+  // Given a call, make a copy of the function it is calling that has more
+  // refined arguments that fit the arguments being passed perfectly.
+  Name getRefinedTarget(Call* call, Module* module) {
+    auto target = call->target;
+    auto* func = module->getFunction(target);
+    if (func->imported()) {
+      // Nothing to do since this calls outside of the module.
+      return target;
+    }
+    auto params = func->getParams();
+    bool hasRefinedParam = false;
+    for (Index i = 0; i < call->operands.size(); i++) {
+      if (call->operands[i]->type != params[i]) {
+        hasRefinedParam = true;
+        break;
+      }
+    }
+    if (!hasRefinedParam) {
+      // Nothing to do since all params are fully refined already.
+      return target;
+    }
+
+    std::vector<Type> refinedTypes;
+    for (auto* operand : call->operands) {
+      refinedTypes.push_back(operand->type);
+    }
+    auto refinedParams = Type(refinedTypes);
+    auto iter = funcParamMap.find({target, refinedParams});
+    if (iter != funcParamMap.end()) {
+      return iter->second;
+    }
+
+    // This is the first time we see this situation. Let's see if it is worth
+    // monomorphizing.
+
+    // Create a new function with refined parameters as a copy of the original.
+    // (Note we must clear stack IR on the original: atm we do not have the
+    // ability to copy stack IR, so we'd hit an internal error. But as we will
+    // be optimizing the function anyhow, we'd be throwing away stack IR later
+    // so this isn't a problem.)
+    func->stackIR.reset();
+    auto refinedTarget = Names::getValidFunctionName(*module, target);
+    auto* refinedFunc = ModuleUtils::copyFunction(func, *module, refinedTarget);
+    TypeUpdating::updateParamTypes(refinedFunc, refinedTypes, *module);
+    refinedFunc->type = HeapType(Signature(refinedParams, func->getResults()));
+
+    // Assume we'll choose to use the refined target, but if we are being
+    // careful then we might change our mind.
+    auto chosenTarget = refinedTarget;
+    if (onlyWhenHelpful) {
+      // Optimize both functions using minimal opts, hopefully enough to see if
+      // there is a benefit to the refined types (such as the new types allowing
+      // a cast to be removed).
+      // TODO: Atm this can be done many times per function as it is once per
+      //       function and per set of types sent to it. Perhaps have some
+      //       total limit to avoid slow runtimes.
+      doMinimalOpts(func);
```

Comment:
It might be worth considering. But the problem is that we don't actually have the property of "already fully optimized", not unless the user built with `--converge`, and not even then since that measures the entire module and not function by function (perhaps we should add `--converge-funcs`..?). In practice it is common that another `-O1` finds a few more bytes to save. And in those cases the refined function would win even if the refined types don't help at all. The only way really avoid that is to do an apples-to-apples comparison.

Reply:
Hmm, in that case can we at least perform the minor optimizations (potentially to a fixed point) on the candidate functions all at once up front so they don't continue changing while we generate monomorphizations? But I still don't quite understand why it's not sufficient to run this optimization late in the user-configured pipeline so that things are "already fully optimized" according to whatever the user's definition of that is.

Reply:
> Hmm, in that case can we at least perform the minor optimizations (potentially to a fixed point) on the candidate functions all at once up front so they don't continue changing while we generate monomorphizations?

Maybe I'm missing your point here. I think you might be concerned about this situation: if `A` and `B` call `C`, then we would optimize `C` for purpose of `A->C` and also separately for `B->C`. So `C` gets optimized twice. That is necessary if the refined types differ, but if they match it's wasteful. However, we won't actually do it twice for matching types: once we see a pair of function and types we compute things once, store it in the cache `funcParamMap`, and reload from there. That includes the decision that it isn't worth monomorphizing - I'll clarify that in a comment.

Or are you worried about the main loop on functions getting to `A`, ending up optimizing `C`, then getting to `C`? That's fine, it will just look at more optimized code then. This is ordered and deterministic.

> But I still don't quite understand why it's not sufficient to run this optimization late in the user-configured pipeline so that things are "already fully optimized" according to whatever the user's definition of that is.

If a target function `T` has cost 100, and each subsequent run of `-O1` subtracts one, then it would take 100 full cycles to get to 0. `--converge` will do that (at least in a module with a single function), but generally we don't.

(And, since we don't, if we compare the cost of `T` to the cost of `T` with some refined types + `-O1` then the latter would look better just due to the opts even if refining doesn't help.)

Reply:
I think @tlively is worried about us ending up optimizing `func` multiple times, namely per every possible `refinedFunc`.

Reply:
I see, thanks @aheejin , sorry for missing that.

I guess we could avoid that duplicated work if we saved the older versions and did a comparison using those, rather than the latest optimized code.

Reply:
> Maybe I'm missing your point here. I think you might be concerned about this situation: if `A` and `B` call `C`, then we would optimize `C` for purpose of `A->C` and also separately for `B->C`. So `C` gets optimized twice.

Right, and the weird part here is that `C` has been optimized once when we decide whether to monomorphize the `A` callsite and twice when we decide whether to monomorphize the `B` callsite, so the two potential monomorphizations are not competing on equal footing. Our monomorphization decisions could change depending on the order in which we visit the `A` and `B` callsites. Yes, it's deterministic, but I would hope that our optimization decisions would be more path-independent. This will be more practically important if we ever try to parallelize some of this work.

Since each function has been optimized an arbitrary number of times and is arbitrarily far away from convergence at the point its cost is measured in the current code, it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic, but it could save us half the optimization work in this pass, and it also solves the path-dependence problem described above.

But I guess you're concerned about comparing the original function with N rounds of optimization to the monomorphized function with N+1 rounds of optimization? That makes sense, and I would solve it by comparing an optimized copy of the original function to a monomorphized + optimized copy of the function. To avoid holding multiple copies of every function at once, the top-level loop could be changed to loop first through callees (i.e. optimization candidates) rather than callers. It's a little more complicated, but I think this would be worth the effort. (It would be more cache-friendly, too.)

Reply:
> it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic

I think false positives are quite costly here, that is, cases where we monomorphize unnecessarily. That bloats the code for no good reason and could well end up making this pass appear unuseful. And that can easily happen each time there is even one tiny bit of optimization potential left in a function, if we wrongly assume it is already maximally optimized. So I think it is important to be 100% precise: only monomorphize when it helps.

> comparing an optimized copy of the original function to a monomorphized + optimized copy of the [original] function.

(`[original]` added by me, is that right?) IIUC it would always compare using the original function as a basis, and we'd immediately throw away any improvements we happen to find? I think that can work. However, throwing away even small improvements seems wasteful... Instead, I am leaning towards this recent TODO:

https://github.com/WebAssembly/binaryen/pull/5238/files#diff-e3faa07d688c017f0e69bcd2c24731585177e0c09d40a2034c1ca0111157f0b9R175-R178

Basically, if we optimize a function's cost and it's actually better, then we keep that improvement, and we'd optimize using that later. More work, but we've not thrown out good results. And, in the more common case, there will be no improvement to the cost, in which case we'd mark the function as "maximally optimized" and save that cost, then just use that saved cost later. Thoughts?

Reply:
> > it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic
> 
> I think false positives are quite costly here, that is, cases where we monomorphize unnecessarily... So I think it is important to be 100% precise: only monomorphize when it helps.

Yes, I agree. The solution I described would still only monomorphize when it helps. But it would do about half as much work to make that determination and the results wouldn't depend on the iteration order so it would be easier to parallelize.
 
> > comparing an optimized copy of the original function to a monomorphized + optimized copy of the [original] function.
> 
> (`[original]` added by me, is that right?) 

Yep 👍

> IIUC it would always compare using the original function as a basis, and we'd immediately throw away any improvements we happen to find? ...

No, we should replace the original function with the optimized copy once we no longer need the original version around to create new monomorphized copies of it.

> ... And, in the more common case, there will be no improvement to the cost, in which case we'd mark the function as "maximally optimized" and save that cost, then just use that saved cost later. Thoughts?

I think the "maximally optimized" idea is a red herring. The current code does not ensure or depend on anything being maximally optimized my suggested changes wouldn't change that. Detecting when a function is maximally optimized and caching its cost could save work, but not as much as my proposed changes would.

Reply:
> No, we should replace the original function with the optimized copy once we no longer need the original version around to create new monomorphized copies of it.

I see. That would only help once, though, since it's possible that we optimize `A` into `A'` which is better, and then `A'` can be further improved again to `A''`. That's probably rare though.

A larger concern with saving the original function is that it takes work, since we couldn't just optimize `A` in-place like we normally do, we'd need to first copy `A` and save that copy. In terms of memory usage, it would be better to avoid that, which the suggestion to note which functions are "maximally optimizable" would do (it just stores a bool+Index per function, not a copy of the body).

> I think the "maximally optimized" idea is a red herring. The current code does not ensure or depend on anything being maximally optimized my suggested changes wouldn't change that. Detecting when a function is maximally optimized and caching its cost could save work, but not as much as my proposed changes would.

Fair points, correct. Noting "maximally optimizable" functions would save some but not all the optimization work. It would save the copying and memory usage, though, and it would allow maximal improvement (`A''` from before).

Reply:
> That would only help once, though, since it's possible that we optimize `A` into `A'` which is better, and then `A'` can be further improved again to `A''`. That's probably rare though.

And it shouldn't be the responsibility of this pass to do this extra optimization if it is possible. Since the current implementation happens to do repeated optimizations, it makes sense to persist those optimizations, but it would be better to prioritize the efficiency of the implementation and let the overall pass pipeline perform any additional optimizations.

> A larger concern with saving the original function is that it takes work, since we couldn't just optimize `A` in-place like we normally do, we'd need to first copy `A` and save that copy. In terms of memory usage...

Copying the function once seems like a trivial amount of work compared to running the full default function optimization pipeline on it. If we take care to only keep one of these function copies in memory at a time, the memory overhead also doesn't seem problematic. I guess with our arena allocation, the memory would never actually be freed, but that could be solved by allocating the copy in a temporary arena. Maybe all the candidate monomorphised functions should be held in temporary arenas until we decide to definitely include them, at which point they can be copied back into the main module?

Reply:
Copying is not much work, agreed. But the memory usage is more significant I think. We could explore ways to reduce the cost but the only reasonable idea I have is to make a new arena for all these functions, which would mean it lives til the end of the pass. That could be a significant amount of memory in some cases.

I agree it shouldn't be the responsibility of this pass to do incidental optimizations. But it is just simpler to do them and keep them.

I agree there is a downside to what I propose, in that it will do somewhat more optimization work than needed. But in the common case (the function is already fully optimized) it will do 0 extra work. And it keeps more incidental optimizations it happens to find.

Reply:
Talked offline and I didn't realize how fast running the O1 optimization pipeline was. With a comment about the path-dependence property, this implementation LGTM for experimentation.

We also discussed some other potential future improvements such as monomorphizing for specific values in addition to refined types.